pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,214 @@
+
+	/* There is some weirdness with diagnostics flag usage. The same flags
+	   object is shared between the (implied) "all" cmd and the client cmd.
+
+	   We actually use the client factory built in the "client" subcommand for
+	   discovery in either case, and that adds flags to the "client" cmd which
+	   need a target for putting results in; we do not want the factory to add
+	   those flags to the "all" command (the only client option there is a
+	   config file), but those flags have to exist somewhere for the factory
+	   to set values and look them up later (even though on the "all" command
+	   the flags will not exist).
+
+	   So the flags object from client cmd is reused for the "all" command.
+	*/
+
+	ccmd, factory := NewClientCommand(name+" client", diagFlags)
+	cmd.AddCommand(ccmd)
+	cmd.AddCommand(NewMasterCommand(name + " master"))
+	cmd.AddCommand(NewNodeCommand(name+" node", diagFlags))
+	cmd.AddCommand(NewOptionsCommand())
+
+	cmd.Run = func(c *cobra.Command, args []string) {
+		runInit(c, diagFlags)
+		diagFlags.CanCheck[types.ClientTarget] = true
+		diagFlags.CanCheck[types.MasterTarget] = true
+		diagFlags.CanCheck[types.NodeTarget] = true
+		run.Diagnose(diagFlags, factory)
+	}
+	return cmd
+}
+
+// Used in the command definition for common flags that we want to be visible in usage
+func addFlags(cmd *cobra.Command, flags *types.Flags) {
+	cmd.Flags().VarP(&flags.Diagnostics, "diagnostics", "d", `comma-separated list of diagnostic names to run, e.g. "systemd.AnalyzeLogs"`)
+	cmd.Flags().IntVarP(&flags.LogLevel, "diaglevel", "l", 3, "Level of output: 0: Error, 1: Warn, 2: Notice, 3: Info, 4: Debug")
+	cmd.Flags().StringVarP(&flags.Format, "output", "o", "text", "Output format: text|json|yaml")
+}
+
+// Every command invocation needs to do the same things at the beginning...
+func runInit(cmd *cobra.Command, diagFlags *types.Flags) {
+	cmd.SetOutput(os.Stdout)                         // for output re usage / help
+	diagFlags.OpenshiftFlags = cmd.PersistentFlags() // capture flags from *this* command in env
+	log.SetLevel(diagFlags.LogLevel)
+	log.SetLogFormat(diagFlags.Format) // note, ignore error; if format unknown, just do text
+}
+
+const longClientDescription = `
+OpenShift Diagnostics
+
+This command helps you understand and troubleshoot OpenShift as a user. It is
+intended to be run from the same context as an OpenShift client
+("openshift cli" or "osc") and with the same configuration options.
+
+    $ %s
+`
+
+func NewClientCommand(fullName string, diagFlags *types.Flags) (*cobra.Command, *osclientcmd.Factory) {
+	cmd := &cobra.Command{
+		Use:   "client",
+		Short: "Troubleshoot using the OpenShift v3 client.",
+		Long:  fmt.Sprintf(longClientDescription, fullName),
+	}
+
+	// Add some diagnostics flags to be shown separately from client flags
+	addFlags(cmd, diagFlags)
+	cmd.Flags().StringVarP(&diagFlags.OpenshiftPath, "openshift", "", "", "Path to 'openshift' binary")
```

Comment:
The idea was that for a side-loaded diagnostics command (which would be likely at least for a while) it would be good to know if there was a difference in versions between the openshift binary in use and the one doing the diagnostics (so if running diagnostics from origin versus an enterprise install a few months back, we would expect different capabilities, or just warn about the difference). Also it would be useful to find in a process list.

But several factors have emerged that probably make this fairly irrelevant now.